rm/rm-client/src/main/java/org/ow2/proactive/resourcemanager/common/event/RMInitialState.java

```diff
@@ -55,67 +58,118 @@
 @XmlRootElement
 public class RMInitialState implements Serializable {
 
-    /** Nodes events */
-    private ArrayList<RMNodeEvent> nodesList = new ArrayList<>();
+    /**
+     * Nodes events
+     */
+    private Map<String, RMNodeEvent> nodeEvents = new ConcurrentHashMap<>();
+
+    /**
+     * Nodes sources AO living in RM
+     */
+    private Map<String, RMNodeSourceEvent> nodeSourceEvents = new ConcurrentHashMap<>();
 
-    /** Nodes sources AO living in RM */
-    private ArrayList<RMNodeSourceEvent> nodeSources = new ArrayList<>();
+    private long latestCounter = -1;
 
     /**
      * ProActive empty constructor
      */
     public RMInitialState() {
+
     }
 
     /**
      * Creates an InitialState object.
      *
-     * @param nodesEventList RM's node events.
+     * @param nodesEventList  RM's node events.
      * @param nodeSourcesList RM's node sources list.
      */
-    public RMInitialState(ArrayList<RMNodeEvent> nodesEventList, ArrayList<RMNodeSourceEvent> nodeSourcesList) {
-        this.nodesList = nodesEventList;
-        this.nodeSources = nodeSourcesList;
+    public RMInitialState(Map<String, RMNodeEvent> nodesEventList, Map<String, RMNodeSourceEvent> nodeSourcesList) {
+        this.nodeEvents = nodesEventList;
+        this.nodeSourceEvents = nodeSourcesList;
     }
 
     /**
-     * Returns the 'to release' Nodes list.
-     * @return the 'to release' Nodes list.
+     * Current version of RM portal and maybe other clients expects "nodesEvents" inside JSON
+     *
+     * @return
      */
-    public ArrayList<RMNodeEvent> getNodesEvents() {
-        return this.nodesList;
+    public List<RMNodeEvent> getNodesEvents() {
+        return new ArrayList(this.nodeEvents.values());
     }
 
     /**
-     * Returns the NodeSources list.
-     * @return the NodeSources list.
+     * Current version of RM portal and maybe other clients expects "nodeSource" inside JSON
+     *
+     * @return
      */
-    public ArrayList<RMNodeSourceEvent> getNodeSource() {
-        return this.nodeSources;
+    public List<RMNodeSourceEvent> getNodeSource() {
+        return new ArrayList<>(this.nodeSourceEvents.values());
     }
 
-    public void nodeStateChanged(RMNodeEvent stateChangedEvent) {
-        int size = nodesList.size();
-        for (int i = 0; i < size; i++) {
-            if (stateChangedEvent.getNodeUrl().equals(nodesList.get(i).getNodeUrl())) {
-                nodesList.set(i, stateChangedEvent);
-                break;
-            }
-        }
+    public long getLatestCounter() {
+        return latestCounter;
+    }
+
+    public void nodeAdded(RMNodeEvent event) {
+        updateCounter(event);
```

Comment:
In general, shouldn't we discard the event if the event's counter is less than the latest counter?

Reply:
But i would prefer to not discard anything - because of the counter we can lost an event. Maybe it is better to log if we have suspicious counter